src/iface/fragments.rs

```diff
@@ -0,0 +1,291 @@
+    pub fn add(&mut self, new_packet: Packet<'a>) {
+        match self.packets {
+            ManagedSlice::Borrowed(ref mut packets) => {
+                for packet in packets.iter_mut() {
+                    match packet.state {
+                        PacketState::Empty => {
+                            *packet = new_packet;
+                            return;
+                        },
+                        PacketState::Assembling{..} => {}
+                    }
+                }
+            },
+            #[cfg(any(feature = "std", feature = "alloc"))]
+            ManagedSlice::Owned(ref mut packets) => {
+                packets.push(new_packet);
+            }
+        }
+    }
+
+    /// Get a packet with given key
+    fn get(&mut self, key_id: u16, key_src_addr: Ipv4Address, key_dst_addr: Ipv4Address) -> Option<&mut Packet<'a>> {
+        for packet in self.packets.iter_mut() {
+            match packet.state {
+                PacketState::Empty => {},
+                PacketState::Assembling{id, src_addr, dst_addr, ..} => {
+                    if key_id == id &&
+                       key_src_addr == src_addr &&
+                       key_dst_addr == dst_addr {
+                           return Some(packet);
+                       }
+                }
+            }
+        }
+        None
+    }
+
+    /// Get an empty packet
+    fn get_empty(&mut self) -> Option<&mut Packet<'a>> {
+        for packet in self.packets.iter_mut() {
+            match packet.state {
+                PacketState::Empty => {return Some(packet);},
+                PacketState::Assembling{..} => {}
+            }
+        }
+        None
+    }
+
+
+    /// Remove stale packets
+    fn purge_old_packets(&mut self, time: Instant) {
+        for packet in self.packets.iter_mut() {
+            match packet.state {
+                PacketState::Empty => {},
+                PacketState::Assembling{last_used, ..} => {
+                    if (time.total_millis() - last_used.millis) > Self::FRAGMENTATION_TIMEOUT_MS {
+                        packet.reset();
+                    }
+                }
+            }
+        }
+    }
+
+    /// Check if the set contains a packet with given `id`
+    pub fn contains(&self, key_id: u16, key_src_addr: Ipv4Address, key_dst_addr: Ipv4Address) -> bool {
+        for packet in self.packets.iter() {
+            match packet.state {
+                PacketState::Empty => {},
+                PacketState::Assembling{id, src_addr, dst_addr, ..}=> {
+                    if key_id == id &&
+                       key_src_addr == src_addr &&
+                       key_dst_addr == dst_addr {
+                           return true;
+                       }
+                },
+            }
+        }
+        false
+    }
+
+    /// Get a packet with given key
+    pub fn get_packet(&mut self, id: u16, src_addr: Ipv4Address, dst_addr: Ipv4Address, time: Instant) -> Option<&mut Packet<'a>> {
```

Comment:
The is `O(3n)` right? Is there a way we could make this faster? What if we used a `ManagedMap` instead of a `ManagedSlice`?

Reply:
There is a thread in the old PR where I tried to explain why `ManagedMap` doesn't work. In short, to insert a new packet to `ManagedMap` you need to allocate memory for it somewhere. And either you create a new `Vec<u8>` which is easy but requires `alloc`. Or you have a static buffers somewhere, and that effectively makes it a `ManagedSlice`.

I am not sure how much fragmented packets you normally encounter (TCP doesn't use fragmented packets typically), in our tests the performance penalty was negligible (with large UDP packets).

But - to use fragmentation in `no-std` builds, you still need to solve problem with `From` conversion for `ManagedSlice` here: https://github.com/m-labs/rust-managed/blob/master/src/slice.rs#L74

If you don't have alloc/std then rust doesn't know how to create a `FragmentSet` from static buffers. It would be cool to be able to do fragmentation without `alloc`, but I am not sure what is the best way around this.

Reply:
Edit: I realized even the `loopback.rs` example you have to run with at least `alloc` because of the `From` trait implementation. This definitely limits smoltcp's usage on bare metal systems:-(

Reply:
> And either you create a new Vec<u8> which is easy but requires alloc. Or you have a static buffers somewhere, and that effectively makes it a ManagedSlice.

Yeah it is basically the same, but you get massivly improved lookup times due to the use of a btree. Couldn't you just use something like the following.

```rust
const FRAGMENT_BUFFER_SIZE: usize = // ...
static FRAGMENT_BUFFER: [Option<Ident, FragmentPacket> ; FRAGMENT_BUFFER_SIZE] = [None; FRAGMENT_BUFFER_SIZE];
let fragment_set = FragmentSet::new(FRAGMENT_BUFFER);
```
> I am not sure how much fragmented packets you normally encounter (TCP doesn't use fragmented packets typically), in our tests the performance penalty was negligible (with large UDP packets).

Lots of embedded applicaitons make heavy use of fragmented packets. See 6LoWPAN ([RFC 4944] and [RFC 6282]). As `smoltcp` matures this is going to become more and more important. IMO the next two big hurdles are this and breaking up the `EthernetInterface`.

[RFC 4944]: https://tools.ietf.org/html/rfc4944
[RFC 6282]: https://tools.ietf.org/html/rfc6282

Reply:
Thanks @dlrobertson, these are my thoughts exactly.